x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizerTests.java

```diff
@@ -3266,6 +3269,83 @@ public void testPlanSanityCheck() throws Exception {
         assertThat(e.getMessage(), containsString(" optimized incorrectly due to missing references [salary"));
     }
 
+    /**
+     * Pushing down EVAL/GROK/DISSECT/ENRICH must not accidentally shadow attributes required by SORT.
+     *
+     * For DISSECT expects the following; the others are similar.
+     *
+     * EsqlProject[[emp_no{r}#29, salary{r}#30]]
+     * \_TopN[[Order[$$emp_no$temp_name$34{r}#44 + $$salary$temp_name$39{r}#45 * 13[INTEGER],ASC,LAST], Order[NEG($$salary$t
+     * emp_name$39{r}#45),DESC,FIRST]],3[INTEGER]]
+     *   \_Dissect[first_name{f}#35,Parser[pattern=%{emp_no} %{salary}, appendSeparator=, parser=org.elasticsearch.dissect.Dissect
+     * Parser@39042612],[emp_no{r}#29, salary{r}#30]]
+     *     \_Eval[[emp_no{f}#34 AS $$emp_no$temp_name$34, salary{f}#39 AS $$salary$temp_name$39]]
+     *       \_EsRelation[test][_meta_field{f}#40, emp_no{f}#34, first_name{f}#35, ..]
+     */
+    public void testPushdownWithOverwrittenName() throws Exception {
+        List<String> overwritingCommands = List.of(
+            "EVAL emp_no = 3*emp_no, salary = -2*emp_no-salary",
+            "DISSECT first_name \"%{emp_no} %{salary}\"",
+            "GROK first_name \"%{WORD:emp_no} %{WORD:salary}\"",
+            "ENRICH languages_idx ON first_name WITH emp_no = language_code, salary = language_code"
+        );
+
+        for (String overwritingCommand : overwritingCommands) {
+            var plan = optimizedPlan(LoggerMessageFormat.format(null, """
+                FROM test
+                | SORT 13*(emp_no+salary) ASC, -salary DESC
+                | {}
+                | LIMIT 3
+                | KEEP emp_no, salary
```

Comment:
It would be nice to have a test with `FROM test | KEEP emp_no, salary | SORT ...` (instead of KEEP at the end) just to make sure that the temporary attribute does not leak at the end.

Reply:
Very good remark, thank you.